Add bedrock2 instructions to README (#770)
mit-plv/fiat-crypto: README.md

Fiat-Crypto: Synthesizing Correct-by-Construction Code for Cryptographic Primitives
===================================================================================

Building
--------
[![CI (Coq)](https://github.com/mit-plv/fiat-crypto/workflows/CI%20(Coq)/badge.svg)](https://github.com/mit-plv/fiat-crypto/actions?query=workflow%3A%22CI+%28Coq%29%22+branch%3Amaster)

This repository requires [Coq](https://coq.inria.fr/) [8.9](https://github.com/coq/coq/releases/tag/V8.9.0) or later.
Note that if you install Coq from Ubuntu aptitude packages, you need `libcoq-ocaml-dev` in addition to `coq`.
If you want to build the bedrock2 code, you need [Coq 8.10](https://github.com/coq/coq/releases/tag/V8.10.0) or later (otherwise you can pass `SKIP_BEDROCK2=1` to `make`).

> We suggest downloading [the latest version of Coq](https://coq.inria.fr/download).

Reviewer: Please link to https://github.com/coq/coq/wiki#coq-installation instead

Author: ```suggestion
We suggest downloading [the latest version of Coq](https://github.com/coq/coq/wiki#coq-installation).
```

You can clone this repository with

    git clone --recursive https://github.com/mit-plv/fiat-crypto.git

Git submodules are used for some dependencies. If you did not clone with `--recursive`, run

    git submodule update --init --recursive

from inside the repository.
Then run:

    make

> You can check out [our CI](https://github.com/mit-plv/fiat-crypto/actions?query=branch%3Amaster+workflow%3A%22CI+%28Coq%29%22) to see how long the build should take; as of the last update to this line in the README, it takes about 1h10m to run `make -j2` on Coq 8.11.1.

Reviewer: I think it is worth mentioning for the order of magnitude (i.e., not seconds) but yeah we have no claim for it being reproducible

> You can check out [our CI](https://github.com/mit-plv/fiat-crypto/actions?query=branch%3Amaster+workflow%3A%22CI+%28Coq%29%22) to see how long the build should take; as of the last update to this line in the README, it takes about 1h10m to run `make -j2` on Coq 8.11.1.

Author: I think it's worth it for calibration.  It's probably not going to finish in 5 minutes regardless of machine, and it's probably not going to take 12 hours, unless the machine is extremely slow.

> You can check out [our CI](https://github.com/mit-plv/fiat-crypto/actions?query=branch%3Amaster+workflow%3A%22CI+%28Coq%29%22) to see how long the build should take; as of the last update to this line in the README, it takes about 1h10m to run `make -j2` on Coq 8.11.1.

Reviewer: I'm fine with it as is. if you want to improve, you could name the machine for which this timing was, even if the best we can do is "on GitHub CI". I think it's better merged than not merged either way.

    make c-files

The C files will appear in [`fiat-c/src/`](./fiat-c/src/).

Just the compilers generating these C files can be made with

    make standalone-ocaml

> or `make standalone-haskell` for binaries generated with Haskell, or `make standalone` for both the Haskell and OCaml binaries.

Reviewer: "compiler binaries"?

Author: ```suggestion
or `make standalone-haskell` for compiler binaries generated with Haskell, or `make standalone` for both the Haskell and OCaml compiler binaries.
```

You can find more examples in the `Makefile`.

Usage (Generating Bedrock2 Files)
---------------------------------

> 

Reviewer 2: ```suggestion
Output is supported to the research language [bedrock2](https://github.com/mit-plv/bedrock2).
```
I'd also like to explain something like "the C code for bedrock2 is really a C wrapper for RISC-V code which is why it's weird" but I'm not sure how to phrase that. @andres-erbsen?

Reviewer: Mh. I am not sure how. Perhaps "the main target of bedrock2 is risc-v; the C code produced by it is unlikely to be idiomatic"?

Author: If one of you wants to make a suggestion, I'm happy to include it.  Otherwise we should probably merge this all soon for @achlipala ?

The Coq development builds binary compilers that generate code using some implementation strategy.
The parameters (modulus, hardware multiplication input bitwidth, etc.) are are specified on the command line of the compiler.

> The generated bedrock2/C code is written to standard output.

Reviewer: Perhaps a sentence to explain that this is fiat-crypto targetting bedrock2, and then printing C code using the bedrock2 C backend?

Author: ```suggestion
The generated bedrock2 code is then written to standard output using the bedrock2 C backend.
```

Author: Is this good?

Reviewer: Yes
